Source/Scene/Scene.js

```diff
@@ -3151,33 +3193,63 @@ define([
         }
     }
 
-    function update(scene) {
+    function prePassesUpdate(scene) {
+        scene._jobScheduler.resetBudgets();
+
         var frameState = scene._frameState;
+        var primitives = scene.primitives;
+        var length = primitives.length;
+        var i;
+        var primitive;
+        for (i = 0; i < length; ++i) {
```

Comment:
I didn't look at the overall logic but is this doing a linear search over all the primitives per frame for 3D Tilesets???

I hope not.

If so, why not check the `instanceof` at `add` time and keep a separate array of them; then just loop through that here.

Also, I do not have time to do a real architecture review, but please ensure that folks are comfortable with how coupled this is now.

Reply:
Even what @pjcozzi is suggesting is flawed because it doesn't account for `Cesium3DTileset` instances that are part of a `PrimitiveCollection`.

I don't know the details, but I'm not sure this is the right approach.  Perhaps a new top-level function for all primitives is needed to avoid having to pluck out 3D Tilesets at all, it would just be a no-op or not implemented by primitives that don't support it.  i.e. anything that implements `prePassesUpdate` gets called and the only two things that implement it is `Cesium3DTileset` and `PrimitiveCollection` for now.

Reply:
Added a function to primitivecollection for the different functions that required doing instanceof. We'll be able to leverage the hierarchy feature of primitivecollections now but still need to perform the instanceof check.  I could add empty function bodies to other primitive types and get rid of the instance check or maintain a 3dtiles collection in primitivecollection.